src/node/coinstats.cpp

```diff
@@ -144,22 +145,32 @@ static bool GetUTXOStats(CCoinsView* view, BlockManager& blockman, CCoinsStats&
     return true;
 }
 
-bool GetUTXOStats(CCoinsView* view, BlockManager& blockman, CCoinsStats& stats, CoinStatsHashType hash_type, const std::function<void()>& interruption_point, const CBlockIndex* pindex, bool index_requested)
+std::optional<CCoinsStats> GetUTXOStats(CCoinsView* view, BlockManager& blockman, CoinStatsHashType hash_type, const std::function<void()>& interruption_point, const CBlockIndex* pindex, bool index_requested)
 {
+    CCoinsStats stats{};
+    bool success;
+
     switch (hash_type) {
     case(CoinStatsHashType::HASH_SERIALIZED): {
         CHashWriter ss(SER_GETHASH, PROTOCOL_VERSION);
-        return GetUTXOStats(view, blockman, stats, ss, interruption_point, pindex, hash_type, index_requested);
+        success = GetUTXOStats(view, blockman, stats, ss, interruption_point, pindex, hash_type, index_requested);
+        break;
     }
     case(CoinStatsHashType::MUHASH): {
         MuHash3072 muhash;
-        return GetUTXOStats(view, blockman, stats, muhash, interruption_point, pindex, hash_type, index_requested);
+        success = GetUTXOStats(view, blockman, stats, muhash, interruption_point, pindex, hash_type, index_requested);
+        break;
     }
     case(CoinStatsHashType::NONE): {
-        return GetUTXOStats(view, blockman, stats, nullptr, interruption_point, pindex, hash_type, index_requested);
+        success = GetUTXOStats(view, blockman, stats, nullptr, interruption_point, pindex, hash_type, index_requested);
+        break;
     }
     } // no default case, so the compiler can warn about missing cases
-    assert(false);
+
+    if (!success) {
```

Comment:
2e1ff3915cab0e902195514f04378f74e3453610 (coinstats: Return purely out-param CCoinsStats)

This gives me a `-Wmaybe-uninitialized` compiler warning for `success` (although this is obviously impossible).

Reply:
Hmmm, do you think we need to do anything here? I'm not sure how best to fix this since we want to keep it so that the compiler can warn about missing cases

Reply:
Maybe just initialise success with `true` or `false` to appease the compiler?
I guess the catch-22 ist that if we had multiple warnings like this on master, we'd easily miss legitimate warnings such as the one about missing cases, which could then lead to `success` actually being uninitialized in this spot...
But I don't feel strongly about this.

Reply:
It is not UB (I think) to pass any underlying integral value as an `enum class`, so uninitialized `success` can in theory happen in a "well-formed" program.

A few alternative fixes:

* Initialize success with `false`
* Keep the `assert(false)` and inline all return statements into the case statments: `if (!Calc(...)) return std::nullopt; return stats;`
* Keep the `assert(false)`, but wrap it with the whole `switch (hash_type)` scope into an inlined lambda
* Something else ... ?

Reply:
In commit "coinstats: Return purely out-param CCoinsStats" (f735c92edc08047bdbe0f47ea0e7c319bcc62eeb)

Agree with Marco, I think. I don't understand why you wouldn't initialize success to false. If function is called with `(CoinStatsHashType)123` compiler is correct to warn that it has undefined behavior. It seems like it would be best to assert false in this case, second best to return nullopt if asserting is inconvenient, and third best to have undefined behavior.

Reply:
Fixed in 524463daf6a10b20a4e20116a68101a684929eda using a lambda, thanks for the help here folks!